org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/parser/CompUtil.java

```diff
@@ -237,7 +237,7 @@ public Object visit(ExprBinary x) throws Err {
      * @param thispath - the set of filenames involved in the current
      *            chain_of_file_opening
      */
-    private static CompModule parseRecursively(List<Object> seenDollar, Map<String,String> loaded, Map<String,String> fc, Pos pos, String filename, CompModule root, String prefix, Set<String> thispath, int initialResolution) throws Err, FileNotFoundException, IOException {
+    private static CompModule parseRecursively(List<Object> seenDollar, Map<String, CompModule> modCache, Map<String,String> loaded, Map<String,String> fc, Pos pos, String filename, CompModule root, String prefix, Set<String> thispath, int initialResolution) throws Err, FileNotFoundException, IOException {
```

Comment:
The modcache and the loaded maps seem to have the same responsibility? I think it would be a lot better if this was abstracted in a separate class. If you do, please make an interface for the API and an implementation.